lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -374,31 +373,28 @@ def write(chunk)
             if record.key?('kubernetes')
               handle_container_metadata(record, entry)
             end
-            # Save the timestamp if available
-            timestamp = record.key?('time') ? record['time'] : nil
-            # If the log from the user container is json, we want to export it
-            # as a structured log. Now that we've pulled out all the
-            # container-specific metadata from the record, we can replace the
-            # record with the json that the user logged.
-            # To save CPU in the common case of unstructured logs, only check if
-            # the contents are parsable as json for the first entry of each
-            # batch.
-            if is_container_json.nil? && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              if record_json.nil?
-                is_container_json = false
-              else
-                record = record_json
-                is_container_json = true
-              end
-            elsif is_container_json && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              record = record_json unless record_json.nil?
-            end
-            # Restore timestamp if necessary
-            unless record.key?('time') || timestamp.nil?
-              record['time'] = timestamp
-            end
+          end
+
+          # Save the timestamp if available, then clear it out to allow for
+          # determining whether we should parse the log or message field.
+          timestamp = record.key?('time') ? record['time'] : nil
+          record.delete('time')
+          # If the log is json, we want to export it as a structured log unless
+          # there is additional metadata that would be lost.
+          is_json = false
+          if record.length == 1 && record.key?('log')
+            record_json = parse_json_or_nil(record['log'])
```

Comment:
I would expect the performance to be better for the normal car (log steams
not containing JSON), because of the simple string prefix check for an
opening curly brace (instead of relying on catching exceptions), and not
worse for those log streams containing only JSON. The only expected
slowdown would be in cases where log lines frequently start with a curly
brace but do not contain valid JSON.

On Wed, 15 Jun 2016, 18:48 igorpeshansky, <notifications@github.com> wrote:

> In lib/fluent/plugin/out_google_cloud.rb
> <https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/77#discussion_r67199936>
> :
>
> > -            end
> > -            # Restore timestamp if necessary
> > -            unless record.key?('time') || timestamp.nil?
> > -              record['time'] = timestamp
> > -            end
> > +          end
> > +
> > +          # Save the timestamp if available, then clear it out to allow for
> > +          # determining whether we should parse the log or message field.
> > +          timestamp = record.key?('time') ? record['time'] : nil
> > +          record.delete('time')
> > +          # If the log is json, we want to export it as a structured log unless
> > +          # there is additional metadata that would be lost.
> > +          is_json = false
> > +          if record.length == 1 && record.key?('log')
> > +            record_json = parse_json_or_nil(record['log'])
>
> @a-robinson <https://github.com/a-robinson> @rbuskens
> <https://github.com/rbuskens> @mr-salty <https://github.com/mr-salty> I'm
> a bit concerned about this now being done unconditionally, outside of the
> CONTAINER_SERVICE branch. We already have complaints about output plugin
> performance, and this is likely to exacerbate that problem... Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/77/files/eef0d044ce874f422dff0aa2e47e51373c3dca9e#r67199936>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALxtGk2TIiCKOOi9-D0tOt0ZWr_Hl0Kks5qMCzFgaJpZM4I1yzk>
> .
>